File: frontend/src/api/portfolio.ts
import http from './http'

export interface OpenTradeItem {
  trade_id: number
  stock_code: string
  stock_name: string | null
  total_qty: number | null
  avg_cost: number | null
  ref_close: number | null
  ref_close_date: string | null
  ref_market_value: number | null
  ref_pnl: number | null
  ref_pnl_pct: number | null
  has_ref_price: boolean
}

export interface ClosedTradeItem {
  trade_id: number
  stock_code: string
  stock_name: string | null
  closed_at: string | null
  realized_pnl: number | null
  realized_pnl_rate: number | null
  review_text: string | null
  image_count: number
}

export interface OperationOut {
  id: number
  op_type: string
  op_date: string
  qty: number
  price: number
  operation_rating: string | null
  note: string | null
}

export interface TradeDetail {
  trade: {
    id: number
    stock_code: string
    stock_name: string | null
    status: string
    opened_at: string | null
    closed_at: string | null
    avg_cost: number | null
    total_qty: number | null
    realized_pnl: number | null
    realized_pnl_rate: number | null
    review_text: string | null
  }
  operations: OperationOut[]
  images: { id: number; url: string }[]
}

export interface StatsResponse {
  stock_win_rate: {
    won: number
    lost: number
    breakeven: number
    total: number
    rate: number | null
  }
  operation_win_rate: {
    good: number
    bad: number
    unrated: number
    rated_total: number
    rate: number | null
  }
  overall_pnl: {
    total_profit: number
    total_loss: number
    net_pnl: number
    net_pnl_rate: number | null
  }
}

export function getOpenTrades() {
  return http.get<{ items: OpenTradeItem[] }>('/portfolio/open-trades')
}

export function openTrade(data: {
  stock_code: string
  op_date: string
  qty: number
  price: number
  fee?: number
}) {
  return http.post<{ trade_id: number; operation_id: number }>('/portfolio/trades/open', data)
}

export function addOperation(
  tradeId: number,
  data: {
    op_type: 'add' | 'reduce'
    op_date: string
    qty: number
    price: number
    fee?: number
    operation_rating?: string | null
    note?: string | null
  }
) {
  return http.post<{ operation_id: number }>(`/portfolio/trades/${tradeId}/operations`, data)
}

export function closeTrade(
  tradeId: number,
  data: {
    op_date: string
    qty: number
    price: number
    fee?: number
    operation_rating?: string | null
    note?: string | null
  }
) {
  return http.post<{ trade_id: number; realized_pnl: number | null }>(
    `/portfolio/trades/${tradeId}/close`,
    data
  )
}

export function deleteOpenTrade(tradeId: number) {
  return http.delete(`/portfolio/trades/${tradeId}`)
}

export function getClosedTrades(params?: { page?: number; page_size?: number; stock_code?: string }) {
  return http.get<{ total: number; items: ClosedTradeItem[] }>('/portfolio/closed-trades', { params })
}

export function getTradeDetail(tradeId: number) {
  return http.get<TradeDetail>(`/portfolio/trades/${tradeId}`)
}

export function patchReview(tradeId: number, review_text: string | null) {
  return http.patch(`/portfolio/trades/${tradeId}/review`, { review_text })
}

export function uploadReviewImage(tradeId: number, file: File) {
  const fd = new FormData()
  fd.append('file', file)
  return http.post<{ image_id: number; url: string }>(`/portfolio/trades/${tradeId}/images`, fd, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export function deleteReviewImage(imageId: number) {
  return http.delete(`/portfolio/images/${imageId}`)
}

export function patchOperationRating(operationId: number, operation_rating: 'good' | 'bad' | null) {
  return http.patch(`/portfolio/operations/${operationId}/rating`, { operation_rating })
}

export function getStats(params?: { from_date?: string; to_date?: string }) {
  return http.get<StatsResponse>('/portfolio/stats', { params })
}

/** 拉取图片为 blob，用于需 Authorization 的预览 */
export async function fetchImageBlob(imageId: number): Promise<string> {
  const res = await http.get(`/portfolio/images/${imageId}/file`, { responseType: 'blob' })
  return URL.createObjectURL(res.data as Blob)
}

File: backend/app/schemas/portfolio.py
"""个人持仓 API 请求/响应模式。"""

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class TradeOpenBody(BaseModel):
    stock_code: str = Field(..., min_length=1, max_length=20)
    op_date: date
    qty: Decimal = Field(..., gt=0)
    price: Decimal = Field(..., gt=0)
    fee: Decimal | None = Field(default=None, ge=0)


class OperationBody(BaseModel):
    op_type: Literal["add", "reduce"]
    op_date: date
    qty: Decimal = Field(..., gt=0)
    price: Decimal = Field(..., gt=0)
    fee: Decimal | None = Field(default=None, ge=0)
    operation_rating: Literal["good", "bad"] | None = None
    note: str | None = None


class CloseBody(BaseModel):
    op_date: date
    qty: Decimal = Field(..., gt=0)
    price: Decimal = Field(..., gt=0)
    fee: Decimal | None = Field(default=None, ge=0)
    operation_rating: Literal["good", "bad"] | None = None
    note: str | None = None


class ReviewPatchBody(BaseModel):
    review_text: str | None = None


class OperationRatingBody(BaseModel):
    operation_rating: Literal["good", "bad"] | None = None


class OpenTradeResponse(BaseModel):
    trade_id: int
    operation_id: int


class OperationCreateResponse(BaseModel):
    operation_id: int


class CloseTradeResponse(BaseModel):
    trade_id: int
    realized_pnl: Decimal | None


class OpenTradeItem(BaseModel):
    trade_id: int
    stock_code: str
    stock_name: str | None
    total_qty: Decimal | None
    avg_cost: Decimal | None
    ref_close: Decimal | None = None
    ref_close_date: date | None = None
    ref_market_value: Decimal | None = None
    ref_pnl: Decimal | None = None
    ref_pnl_pct: Decimal | None = None
    has_ref_price: bool = False


class OpenTradesResponse(BaseModel):
    items: list[OpenTradeItem]


class ClosedTradeItem(BaseModel):
    trade_id: int
    stock_code: str
    stock_name: str | None
    closed_at: datetime | None
    realized_pnl: Decimal | None
    realized_pnl_rate: Decimal | None = None
    review_text: str | None
    image_count: int = 0


class ClosedTradesResponse(BaseModel):
    total: int
    items: list[ClosedTradeItem]


class TradeOut(BaseModel):
    id: int
    stock_code: str
    stock_name: str | None = None
    status: str
    opened_at: datetime | None = None
    closed_at: datetime | None = None
    avg_cost: Decimal | None = None
    total_qty: Decimal | None = None
    realized_pnl: Decimal | None = None
    realized_pnl_rate: Decimal | None = None
    review_text: str | None = None


class OperationOut(BaseModel):
    id: int
    op_type: str
    op_date: date
    qty: Decimal
    price: Decimal
    operation_rating: str | None = None
    note: str | None = None


class TradeImageOut(BaseModel):
    id: int
    url: str


class TradeDetailResponse(BaseModel):
    trade: TradeOut
    operations: list[OperationOut]
    images: list[TradeImageOut] = []


class ImageUploadResponse(BaseModel):
    image_id: int
    url: str


class StockWinRate(BaseModel):
    won: int
    lost: int
    breakeven: int
    total: int
    rate: float | None


class OperationWinRate(BaseModel):
    good: int
    bad: int
    unrated: int
    rated_total: int
    rate: float | None


class OverallPnlSummary(BaseModel):
    total_profit: Decimal
    total_loss: Decimal
    net_pnl: Decimal
    net_pnl_rate: float | None


class StatsResponse(BaseModel):
    stock_win_rate: StockWinRate
    operation_win_rate: OperationWinRate
    overall_pnl: OverallPnlSummary
